File: behavior_tree.py
#coding=utf-8
import xml.etree.ElementTree as ET
import sys

TYPES_ = {
    'int': int,
    'float': float,
    'str': unicode,
}

def make_root(root):
    pass

def make_selector(node):
    funcs = [parse_node(child) for child in node[0]]
    def selector(robot):
        for func in funcs:
            result = func(robot)
            if result:
                return True
        return False
    return selector

def make_sequence(node):
    funcs = [parse_node(child) for child in node[0]]
    def sequence(robot):
        for func in funcs:
            result = func(robot)
            if not result:
                return False
        return True
    return sequence

def make_parallel(node):
    funcs = [parse_node(child) for child in node[0]]
    def parallel(robot):
        results = []
        for func in funcs:
            result = func(robot)
            results.append(result)
        return any(results)
    return parallel

def make_alternate(node):
    funcs = [parse_node(child) for child in node[0]]
    def alternate(robot):
        if not hasattr(robot, '_alternate_i'):
            robot._alternate_i = 0
        i = robot._alternate_i
        func = funcs[i]
        i += 1
        if i >= len(funcs):
            i = 0
        robot._alternate_i = i
        return func(robot)
    return alternate

def extract_keywords(node):
    attrib = node.attrib
    keywords = {}
    for key, value in attrib.iteritems():
        if key == 'Class':
            continue
        if key.startswith('_'):
            continue
        type_key_name = '_' + key + 'Type'
        type_name = attrib[type_key_name]
        type_ = TYPES_[type_name]
        keywords[key.lower()]  = type_(value)
    return keywords

def make_condition(node, name):
    keywords = extract_keywords(node)
    def condition(robot):
        method = getattr(robot, name)
        return method(**keywords)
    condition.func_name = name
    return condition

def make_action(node, name):
    keywords = extract_keywords(node)
    def action(robot):
        method = getattr(robot, name)
        try:
            return method(**keywords)
        except Exception, e:
            print >> sys.stderr, e
            return False
    action.func_name = name
    return action


def parse_composites(node, name):
    if name == 'Selector':
        return make_selector(node)
    if name == 'Sequence':
        return make_sequence(node)
    if name == 'Parallel':
        return make_parallel(node)
    if name == 'Alternate':
        return make_alternate(node)
    raise Exception('Unknown Composite Node', name)

def make_loop(node):
    child = node[0][0]
    func = parse_node(child)
    count = int(node.attrib['Count'])
    def loop(robot):
        return all(func(robot) for i in xrange(count))
    return loop

def parse_decorators(node, name):
    if name == 'Loop':
        return make_loop(node)
    raise Exception('Unknown Decorator Node', name)

def parse_node(node):
    class_name = node.attrib['Class']
    _, node_type, node_name = class_name.split('.')
    if node_type == 'Composites':
        return parse_composites(node, node_name)
    if node_type == 'Decorators':
        return parse_decorators(node, node_name)
    if node_type == 'Conditions':
        return make_condition(node, node_name)
    if node_type == 'Actions':
        return make_action(node, node_name)
    raise Exception('Unknown Type Node', node_type)

def parse_xml(path):
    tree = ET.parse(path)
    root = tree.getroot()
    root = root[0]
    return parse_node(root[0][0])


if __name__ == '__main__':
    root = parse_xml('test.xml')
    import functools
    class Robot(object):
        def __getattr__(self, name):
            print name
            return functools.partial(lambda self: True, self)
    robot = Robot()
    root(robot)

File: behavior_tree.lua
module(... or 'behavior_tree', package.seeall)
require('LuaXml')

TYPES_ = {
    int = tonumber,
    float = tonumber,
    str = tostring,
}

local function split (str, pat)
    local t = {}  -- NOTE: use {n = 0} in Lua-5.0
    local fpat = '(.-)' .. pat
    local last_end = 1
    local s, e, cap = str:find(fpat, 1)
    while s do
        if s ~= 1 or cap ~= '' then
            table.insert(t, cap)
        end
        last_end = e + 1
        s, e, cap = str:find(fpat, last_end)
    end
    if last_end <= #str then
        cap = str:sub(last_end)
        table.insert(t, cap)
    end
    return t
end

function parse_child_nodes (node)
    local funcs = {}
    for i, child in ipairs(node[1]) do
        funcs[i] = parse_node(child)
    end
    return funcs
end

function make_selector (node)
    local funcs = parse_child_nodes(node)
    return function (robot)
        for func in funcs do
            result = func(robot)
            if result then
                return true
            end
        end
        return false
    end
end

function make_sequence (node)
    local funcs = parse_child_nodes(node)
    return function (robot)
        for func in funcs do
            result = func(robot)
            if not result then
                return false
            end
        end
        return true
    end
end

function make_parallel (node)
    local funcs = parse_child_nodes(node)
    return function (robot)
        local results = {}
        for i, func in ipairs(funcs) do
            results[i] = func(robot)
        end
        for i, result in ipairs(results) do
            if not result then
                return false
            end
            return true
        end
    end
end

function make_alternate (node)
    local funcs = parse_child_nodes(node)
    return function (robot)
        robot._alternate_i = robot._alternate_i or 1
        local i = robot._alternate_i
        local func = funcs[i]
        i = i + 1
        if i > #funcs then
            i = 1
        end
        robot._alternate_i = i
        return func(robot)
    end
end

function parse_composites (node, name)
    if name == 'Selector' then
        return make_selector(node)
    elseif name == 'Sequence' then
        return make_sequence(node)
    elseif name == 'Parallel' then
        return make_parallel(node)
    elseif name == 'Alternate' then
        return make_alternate(node)
    end
    error('Unknown Composite Node' .. name)
end

function make_loop (node)
    local child = node[1][1]
    local func = parse_node(child)
    local count = tonumber(node.Count)
    return function (robot)
        for i=1,count do
            if not func(rebot) then
                return false
            end
        end
        return true
    end
end

function parse_decorators (node, name)
    if name == 'Loop' then
        return make_loop(node)
    end
    error('Unknown Decorator Node' .. name)
end

function extract_keywords (node)
    local keywords = {}
    for key in pairs(node) do
        if (type(key) ~= 'number' and
                key ~= 'Class' and
                key:sub(1, 1) ~= '_') then
            local type_key_name = '_' .. key .. '_type'
            local type_name = node[type_key_name]
            local type_ = TYPES_[type_name]
            keywords[key]  = type_(node[key])
        end
    end
    return keywords
end

function make_condition (node, name)
    local keywords = extract_keywords(node)
    return function (robot)
        local method = robot[name]
        return method(keywords)
    end
end

function make_action (node, name)
    local keywords = extract_keywords(node)
    return function (robot)
        local method = getattr(robot, name)
        local result = pcall(method, keywords)
        if result[1] then
            return true
        else
            io.stderr:write(result[2])
            return false
        end
    end
end

function parse_node (node)
    --print(node.Class)
    local _, node_type, node_name = unpack(split(node.Class, '%.'))
    if node_type == 'Composites' then
        return parse_composites(node, node_name)
    elseif node_type == 'Decorators' then
        return parse_decorators(node, node_name)
    elseif node_type == 'Conditions' then
        return make_condition(node, node_name)
    elseif node_type == 'Actions' then
        return make_action(node, node_name)
    end
    error('Unknown Type Node' .. node_type)
end

function parse_xml (path)
    local root = xml.load(path)
    local node = root[1][1][1]
    return parse_node(node)
end


local what = debug.getinfo(1, 'S').what
if what == 'main' then
    parse_xml('test.xml')
end


